dlib/math/windows.h

```diff
@@ -0,0 +1,264 @@
+               0.1365995 * std::cos(2*phi) -
+               0.0106411 * std::cos(3*phi);
+    }
+
+    // ----------------------------------------------------------------------------------------
+
+    template<typename R>
+    R blackman_harris(std::size_t i, std::size_t N, WindowSymmetry type)
+    /*!
+        This computes the Blackman-Harris window function.
+        See https://en.wikipedia.org/wiki/Window_function#Blackman%E2%80%93Harris_window.
+
+        requires
+            - R is float, double, or long double
+            - 0 <= i < N
+    !*/
+    {
+        static_assert(std::is_floating_point<R>::value, "template parameter must be a floating point type");
+        DLIB_ASSERT(i < N, "index out of range");
+        const std::size_t size = type == SYMMETRIC ? N-1 : N;
+        const R phi = (2.0 * pi * i) / size;
+        return 0.35875 -
+               0.48829 * std::cos(phi) +
+               0.14128 * std::cos(2*phi) -
+               0.01168 * std::cos(3*phi);
+    }
+
+    // ----------------------------------------------------------------------------------------
+
+    template<typename R>
+    R blackman_harris7(std::size_t i, std::size_t N, WindowSymmetry type)
+    /*!
+        This computes the 7-order Blackman-Harris window function.
+
+        requires
+            - R is float, double, or long double
+            - 0 <= i < N
+    !*/
+    {
+        static_assert(std::is_floating_point<R>::value, "template parameter must be a floating point type");
+        DLIB_ASSERT(i < N, "index out of range");
+        const std::size_t size = type == SYMMETRIC ? N-1 : N;
+        const R phi = (2.0 * pi * i) / size;
+        return 0.27105 -
+               0.43329 * std::cos(phi) +
+               0.21812 * std::cos(2*phi) -
+               0.06592 * std::cos(3*phi) +
+               0.01081 * std::cos(4*phi) -
+               0.00077 * std::cos(5*phi) +
+               0.00001 * std::cos(6*phi);
+    }
+
+    // ----------------------------------------------------------------------------------------
+
+    enum WindowType
+    {
+        HANN,
+        BLACKMAN,
+        BLACKMAN_NUTTALL,
+        BLACKMAN_HARRIS,
+        BLACKMAN_HARRIS7,
+        KAISER
+        //GAUSSIAN
+    };
+
+    struct window_args
+    {
+        beta_t beta;
+        //sigma_t sigma; //for GAUSSIAN
+    };
+
+    template<typename R>
+    R window(std::size_t i, std::size_t N, WindowType w, WindowSymmetry type, window_args args)
+    {
+        switch(w)
+        {
+            case HANN:              return hann<R>(i, N, type);
+            case BLACKMAN:          return blackman<R>(i, N, type);
+            case BLACKMAN_NUTTALL:  return blackman_nuttall<R>(i, N, type);
+            case BLACKMAN_HARRIS:   return blackman_harris<R>(i, N, type);
+            case BLACKMAN_HARRIS7:  return blackman_harris7<R>(i, N, type);
+            case KAISER:            return kaiser<R>(i, N, args.beta, type);
+        }
+        return R{};
```

Comment:
Although I'm not super sold on this `window()` function.  Isn't is an unnecessary level of indirection?  Like why can't code just call the function they want to call directly?  That will avoid any runtime expense from a `switch` and also makes the interface more clear since there aren't arguments that are sometimes unused and sometimes applicable.

Reply:
I mention it below, but it's need by `stft()`, `istft()` and the rest.

Reply:
Much easier to tell the STFT functions which window to use via an enum

Reply:
This is adding complication though.  And for instance, the Kaiser beta is just defaulted in the use of this everywhere.  Which makes sense, since you don't want to force `window_args` into the API of all the functions that use windows.

The window switch statement stuff could be avoided by having methods that make and return window lambdas.  E.g.
```
auto kaiser_window(window_symmetry type, beta_t beta) {
    return [=](std::size_t i, std::size_t N) { return kaiser(i, N, type, beta); }
```
Then instead of passing the enum values to functions you pass the output of `kaiser_window()`.  This way the functions that build on top of window functions don't need to know anything about the parameters of the window functions.  It ought to be easier for the compiler to optimize too.  Like it would not surprise me if that `switch` statement resulted in slower execution.

More generally though, why not have functions that just build the window?  Like you call `kaiser_window(N, SYMMETRIC, beta)` and it returns a `dlib::matrix` holding the values.  Isn't that going to be way faster?  The matrix expression stuff will result in the values of the window function being recomputed over and over and they are expensive to compute.

Reply:
I love your idea of passing a lambda! I'll do that later

Reply:
I'm not using a lambda, but a struct with a templated `operator()` function. Otherwise you would have to template the function `kaiser_window()`. Bit overkill. If dlib used  C++14 or above, I could return a generic lambda...

Reply:
Oh right, yeah, c++14.  Soon :)   

Are you still stuck with C++11 compilers?  I should ask around and see how close we are to being able to require C++14.  Maybe now is the time to require it.

Reply:
I'm not. I can use c++14 everywhere and I'm pretty sure c++17 too. I would upgrade to 17 personally. I bet a whole bunch of stuff would get a lot simpler, compile faster and no doubt run faster.

Reply:
I imagine a lot of the meta programming in dlib would get a lot better

Reply:
Right. I'll ask around a bit and see.  C++17 is too new, but I think the time for C++14 is pretty much now.

Reply:
Yeah, I'm calling it now.  C++14 for dlib.  I need to update the CI, some cmake stuff a little, and the docs.  But C++14 is required for the next dlib release :)

Reply:
Nice, you're gonna see a lot of PRs relating to modernising dlib. Generic lambdas here we come!

Reply:
I wonder if a lot of the cmake stuff would get simpler if dlib used a more modern cmake version like 3.10 or above

Reply:
Yeah it would.  I am hesitant to mess with what isn't broken though.  And most of the complication has nothing to do with cmake versions and is just the accumulation of all the little exceptions to people's broken systems I've accumulated over the years.